core/predicate_check.go

```diff
@@ -18,50 +18,90 @@ import (
-	// Check that the transaction can cover its IntrinsicGas (including the gas required by the predicate) before
-	// verifying the predicate.
-	intrinsicGas, err := IntrinsicGas(tx.Data(), tx.AccessList(), tx.To() == nil, rules)
+// CheckBlockPredicates verifies the predicates of a block of transactions and
+// returns the results.
+//
+// Returning an error invalidates the block.
+func CheckBlockPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	txs []*types.Transaction,
+) (predicate.BlockResults, error) {
+	var results predicate.BlockResults
+	// TODO: Calculate tx predicates concurrently.
+	for _, tx := range txs {
+		txResults, err := CheckTxPredicates(rules, predicateContext, tx)
+		if err != nil {
+			return nil, err
+		}
+		results.Set(tx.Hash(), txResults)
+	}
+	return results, nil
+}
+
+// CheckTxPredicates verifies the predicates of a transaction and returns the
+// results.
+//
+// Returning an error invalidates the transaction.
+func CheckTxPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	tx *types.Transaction,
+) (predicate.PrecompileResults, error) {
+	// Check that the transaction can cover its IntrinsicGas, including the gas
+	// required by the predicate, before verifying the predicate.
+	accessList := tx.AccessList()
+	intrinsicGas, err := IntrinsicGas(tx.Data(), accessList, tx.To() == nil, rules)
 	if err != nil {
 		return nil, err
 	}
-	if tx.Gas() < intrinsicGas {
-		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)", ErrIntrinsicGas, tx.Gas(), intrinsicGas)
+	if gas := tx.Gas(); gas < intrinsicGas {
+		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)",
+			ErrIntrinsicGas,
+			gas,
+			intrinsicGas,
+		)
 	}
 
 	rulesExtra := params.GetRulesExtra(rules)
-	predicateResults := make(predicate.PrecompileResults)
 	// Short circuit early if there are no precompile predicates to verify
 	if !rulesExtra.PredicatersExist() {
-		return predicateResults, nil
+		return nil, nil
 	}
 
 	// Prepare the predicate storage slots from the transaction's access list
-	predicateArguments := predicate.FromAccessList(rulesExtra, tx.AccessList())
+	predicateArguments := predicate.FromAccessList(rulesExtra, accessList)
 
-	// If there are no predicates to verify, return early and skip requiring the proposervm block
-	// context to be populated.
+	// If there are no predicates to verify, return early and skip requiring the
+	// proposervm block context to be populated.
 	if len(predicateArguments) == 0 {
-		return predicateResults, nil
+		return nil, nil
 	}
 
 	if predicateContext == nil || predicateContext.ProposerVMBlockCtx == nil {
 		return nil, ErrMissingPredicateContext
 	}
 
+	var (
+		txHash           = tx.Hash()
+		predicateResults = make(predicate.PrecompileResults, len(predicateArguments))
```

Comment:
Since we deferred the allocation of the map, now we can allocate the correct size for it 🚀